logsrvd/sendlog.c

```diff
@@ -1541,43 +1589,63 @@ main(int argc, char *argv[])
     if ((log_info = iolog_parse_loginfo(fp, iolog_dir)) == NULL)
 	goto bad;
 
-    /* Open the I/O log files and seek to restart point if there is one. */
-    if (!iolog_open_all(iolog_dir_fd, iolog_dir, iolog_files, open_mode))
-	goto bad;
-    if (sudo_timespecisset(&restart)) {
-	if (!iolog_seekto(iolog_dir_fd, iolog_dir, iolog_files, &elapsed,
-		&restart))
-	    goto bad;
-    }
-
-    /* Connect to server, setup events. */
-    sock = connect_server(host, port);
-    if (sock == -1)
-	goto bad;
-
-    printf("Connected to %s:%s\n", host, port);
-
     if ((evbase = sudo_ev_base_alloc()) == NULL)
 	sudo_fatal(NULL);
     sudo_ev_base_setdef(evbase);
 
-    if (!client_closure_fill(&closure, sock, &elapsed, &restart, iolog_id, log_info))
-	goto bad;
+    if (testrun)
+        printf("connecting clients...\n");
+
+    for (int i = 0; i < nr_of_conns; i++) {
+        sock = connect_server(host, port);
+        if (sock == -1)
+            goto bad;
+        
+        if (!testrun)
+            printf("Connected to %s:%s\n", host, port);
+
+        closure = client_closure_alloc(sock, &elapsed, &restart, iolog_id, log_info);
+        if (!closure)
+            goto bad;
+
+        /* Open the I/O log files and seek to restart point if there is one. */
+        if (!iolog_open_all(iolog_dir_fd, iolog_dir, closure->iolog_files, open_mode))
+            goto bad;
+        if (sudo_timespecisset(&restart)) {
+            if (!iolog_seekto(iolog_dir_fd, iolog_dir, closure->iolog_files, &elapsed,
+                &restart))
+                goto bad;
+        }
+
+        if (sudo_ev_add(evbase, closure->read_ev, NULL, false) == -1)
+            goto bad;
+    }  
+
+    if (testrun)
+        printf("sending logs...\n");
+
+    struct timeval t_start, t_end, t_result;
+    gettimeofday(&t_start, NULL);
 
-    /* Add read event for the server hello message and enter event loop. */
-    if (sudo_ev_add(evbase, closure.read_ev, NULL, false) == -1)
-	goto bad;
     sudo_ev_dispatch(evbase);
 
-    if (closure.state != FINISHED) {
-	sudo_warnx(U_("exited prematurely with state %d"), closure.state);
-	sudo_warnx(U_("elapsed time sent to server [%lld, %ld]"),
-	    (long long)closure.elapsed->tv_sec, closure.elapsed->tv_nsec);
-	sudo_warnx(U_("commit point received from server [%lld, %ld]"),
-	    (long long)closure.committed.tv_sec, closure.committed.tv_nsec);
-	goto bad;
+    gettimeofday(&t_end, NULL);
+    timersub(&t_end, &t_start, &t_result);
```

Comment:
Not all systems have the timersub macro; please use sudo_timevalsub instead.  Alternately, you could use sudo_gettime_real() to get the time as a timespec and use sudo_timespecsub.